File: frontend/src/modules/app/api/settings/services/service-service.ts
import { Service } from '@/modules/app/api/settings/entities/service';
import { CreateService } from '@/modules/app/api/settings/data-transfer-objects/create-service';
import { UpdateService } from '@/modules/app/api/settings/data-transfer-objects/update-service';
import { http } from '@/core';
import { ServiceConfiguration } from '@/modules/app/api/settings/entities/service-configuration';

// Don't think about the name to hard.
export class ServiceService {
    async getServices(): Promise<Service[]> {
        try {
            const res = await http.get('settings/service');

            if (res.data == null) {
                return [];
            }

            return res.data.map((s: any) => {
                const service = new Service(
                    s.name,
                    s.description,
                    s.pricingMethod,
                    s.configurations.map((c: any) => new ServiceConfiguration(c.vehicleCategoryId, c.price, c.duration))
                );
                service.id = s.id;

                return service;
            });
        } catch (e) {
            if (e.response == null || e.response.status == 404) {
                return [];
            } else {
                throw e;
            }
        }
    }

    async createService(createService: CreateService): Promise<Service> {
        const res = await http.post('settings/service', createService);
        const s = new Service(
            res.data.name,
            res.data.description,
            res.data.pricingMethod,
            res.data.configurations.map((c: any) => new ServiceConfiguration(c.vehicleCategoryId, c.price, c.duration))
        );

        s.id = res.data.id;
        return s;
    }

    async updateService(updateService: UpdateService): Promise<Service> {
        const res = await http.put(`settings/service/${updateService.id}`, updateService);
        const s = new Service(
            res.data.name,
            res.data.description,
            res.data.pricingMethod,
            res.data.configurations.map((c: any) => new ServiceConfiguration(c.vehicleCategoryId, c.price, c.duration))
        );
        s.id = res.data.id;
        return s;
    }

    async deleteService(id: string) {
        await http.delete(`settings/service/${id}`);
    }
}

File: api/src/Domain/Settings/Services/ServiceService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace DetailingArsenal.Domain.Settings {
    public interface IServiceService : IService {
        Task<Service?> FindById(Guid id);
        Task<List<Service>> FindByUser(User user);
        Task<Service> Create(CreateService create, User user);
        Task Update(Service service, UpdateService update);
        Task Delete(Service service);
    }

    public class ServiceService : IServiceService {
        IServiceRepo repo;
        ServiceNameUniqueSpecification uniqueNameSpec;
        ServiceNotInUseSpecification notInUseSpec;

        public ServiceService(IServiceRepo repo, ServiceNameUniqueSpecification uniqueNameSpec, ServiceNotInUseSpecification notInUseSpec) {
            this.repo = repo;
            this.uniqueNameSpec = uniqueNameSpec;
            this.notInUseSpec = notInUseSpec;
        }

        public async Task<Service?> FindById(Guid id) {
            return await repo.FindById(id);
        }

        public async Task<List<Service>> FindByUser(User user) {
            return await repo.FindByUser(user);
        }

        public async Task<Service> Create(CreateService create, User user) {
            var service = Service.Create(
                user!.Id,
                create.Name,
                create.Description,
                create.PricingMethod
            );

            service.Configurations = create.Configurations.Select(
                c => ServiceConfiguration.Create(
                    service.Id,
                    c.VehicleCategoryId,
                    c.Price,
                    c.Duration
                )
            ).ToList();

            await uniqueNameSpec.CheckAndThrow(service);
            await repo.Add(service);

            return service;

        }
        public async Task Update(Service service, UpdateService update) {
            service.Name = update.Name;
            service.Description = update.Description;
            service.PricingMethod = update.PricingMethod;

            service.Configurations = update.Configurations.Select(
                c => ServiceConfiguration.Create(
                    service.Id,
                    c.VehicleCategoryId,
                    c.Price,
                    c.Duration
                )
            ).ToList();

            await uniqueNameSpec.CheckAndThrow(service);
            await repo.Update(service);
        }

        public async Task Delete(Service service) {
            await notInUseSpec.CheckAndThrow(service);
            await repo.Delete(service);
        }
    }
}